File: global-apps/global-chat/front/src/modules/chatroom/ot/ChatRoomOTSystem.js
import {OTSystemBuilder, TransformResult} from 'ot-core';
import {isEqual, intersection} from 'lodash';
import MessageOperation from './MessageOperation';
import CallOperation from './CallOperation';
import DropCallOperation from './DropCallOperation';
import HandleCallOperation from './HandleCallOperation';

const chatRoomOTSystem = new OTSystemBuilder()
  .withEmptyPredicate(MessageOperation, operation => operation.isEmpty())
  .withEmptyPredicate(CallOperation, operation => operation.isEmpty())
  .withEmptyPredicate(DropCallOperation, operation => operation.isEmpty())
  .withEmptyPredicate(HandleCallOperation, operation => operation.isEmpty())

  .withInvertFunction(MessageOperation, operation => operation.invert())
  .withInvertFunction(CallOperation, operation => operation.invert())
  .withInvertFunction(DropCallOperation, operation => operation.invert())
  .withInvertFunction(HandleCallOperation, operation => operation.invert())

  .withTransformFunction(MessageOperation, MessageOperation, (operationLeft, operationRight) => {
    return TransformResult.of([operationRight], [operationLeft]);
  })
  .withTransformFunction(MessageOperation, CallOperation, (operationLeft, operationRight) => {
    return TransformResult.of([operationRight], [operationLeft]);
  })
  .withTransformFunction(MessageOperation, DropCallOperation, (operationLeft, operationRight) => {
    return TransformResult.of([operationRight], [operationLeft]);
  })
  .withTransformFunction(MessageOperation, HandleCallOperation, (operationLeft, operationRight) => {
    return TransformResult.of([operationRight], [operationLeft]);
  })

  .withTransformFunction(CallOperation, CallOperation, (operationLeft, operationRight) => {
    if (!isEqual(operationLeft.getPrev(), operationRight.getPrev())) {
      throw new Error('Previous values should be equal');
    }

    const operationLeftNextValue = operationLeft.getNext();
    const operationRightNextValue = operationRight.getNext();

    if (isEqual(operationLeftNextValue, operationRightNextValue)) {
      return TransformResult.empty();
    }

    if (operationLeftNextValue === null) {
      return TransformResult.left([operationLeft.invert(), operationRight]);
    }

    if (operationRightNextValue === null) {
      return TransformResult.right([operationRight.invert(), operationLeft]);
    }

    if (operationLeftNextValue.timestamp >= operationRightNextValue.timestamp) {
      return TransformResult.right([operationRight.invert(), operationLeft]);
    } else {
      return TransformResult.left([operationLeft.invert(), operationRight]);
    }
  })
  .withTransformFunction(CallOperation, DropCallOperation, (operationLeft, operationRight) => {
    if (operationLeft.getNext() !== null) {
      return TransformResult.right([operationRight.invert(), operationLeft]);
    }

    if (!operationRight.isInvert()) {
      return TransformResult.left([operationLeft.invert(), operationRight]);
    }

    throw new Error('OTTransformException');
  })
  .withTransformFunction(CallOperation, HandleCallOperation, (operationLeft, operationRight) => {
    if (operationLeft.getNext() === null) {
      return TransformResult.right([operationRight.invert(), operationLeft]);
    }

    return TransformResult.of([operationRight], [operationLeft]);
  })

  .withTransformFunction(DropCallOperation, DropCallOperation, (operationLeft, operationRight) => {
    if (operationLeft.isEqual(operationRight)) {
      return TransformResult.empty();
    }

    if (!operationLeft.isInvert() && !operationRight.isInvert()) {
      if (operationLeft.getDropTime() <= operationRight.getDropTime()) {
        return TransformResult.right([operationRight.invert(), operationLeft]);
      } else {
        return TransformResult.left([operationLeft.invert(), operationRight]);
      }
    }

    throw new Error('OTTransformException');
  })
  .withTransformFunction(DropCallOperation, HandleCallOperation, (operationLeft, operationRight) => {
    if (!operationLeft.isInvert()) {
      const handled = operationLeft.getHandled();

      for (const [publicKey, status] of operationRight.getMapOperation().entries()) {
        if (status.next === null) {
          handled.delete(publicKey);
        } else {
          handled.set(publicKey, status.next);
        }
      }

      return TransformResult.right([DropCallOperation.create(
        operationLeft.getPubKey(),
        operationLeft.getPeerId(),
        operationLeft.getTimestamp(),
        handled,
        operationLeft.getDropTime()
      )])
    }

    throw new Error('OTTransformException');
  })

  .withTransformFunction(HandleCallOperation, HandleCallOperation, (operationLeft, operationRight) => {
    const mapOperationLeft = operationLeft.getMapOperation();
    const mapOperationRight = operationRight.getMapOperation();

    if (isEqual(mapOperationLeft, mapOperationRight)) {
      return TransformResult.empty();
    }

    const keysIntersection = intersection([...mapOperationLeft.keys()], [...mapOperationRight.keys()]);

    if (!keysIntersection.length) {
      return TransformResult.of([operationRight], [operationLeft]);
    }

    const rightTransformed = new Map();
    const leftTransformed = new Map();

    for (const [publicKey, status] of mapOperationLeft) {
      if (!keysIntersection.includes(publicKey)) {
        rightTransformed.set(publicKey, status);
      }
    }

    for (const [publicKey, status] of mapOperationRight) {
      if (!keysIntersection.includes(publicKey)) {
        leftTransformed.set(publicKey, status);
      }
    }

    for (const publicKey of keysIntersection) {
      const leftValue = mapOperationLeft.get(publicKey);
      const rightValue = mapOperationRight.get(publicKey);
      const leftValueNext = leftValue.next;
      const rightValueNext = rightValue.next;

      if (leftValueNext === null) {
        leftTransformed.set(publicKey, {
          prev: null,
          next: rightValueNext
        });
        continue;
      }

      if (rightValueNext === null) {
        rightTransformed.set(publicKey, {
          prev: null,
          next: leftValueNext
        });
        continue;
      }

      if (leftValueNext === rightValueNext) {
        continue;
      }

      if (leftValueNext === true) {
        rightTransformed.set(publicKey, {
          prev: rightValueNext,
          next: leftValueNext
        });
      } else {
        leftTransformed.set(publicKey, {
          prev: leftValueNext,
          next: rightValueNext
        });
      }
    }

    return TransformResult.of([new HandleCallOperation(leftTransformed)], [new HandleCallOperation(rightTransformed)]);
  })

  .withSquashFunction(MessageOperation, MessageOperation, doSquash((operationLeft, operationRight) => {
    if (operationRight.isEqual(operationLeft.invert())) {
      return MessageOperation.EMPTY;
    }

    return null;
  }))
  .withSquashFunction(CallOperation, CallOperation, doSquash((operationLeft, operationRight) => {
    if (operationLeft.getPrev() !== null && operationLeft.getNext() !== null && operationRight.getPrev() !== null &&
      operationRight.getNext() !== null) {
      return new CallOperation(operationLeft.getPrev(), operationRight.getNext());
    }

    return null;
  }))
  .withSquashFunction(DropCallOperation, DropCallOperation, doSquash((operationLeft, operationRight) => {
    if (operationLeft.isInversionFor(operationRight)) {
      return DropCallOperation.EMPTY;
    }

    return null;
  }))
  .withSquashFunction(HandleCallOperation, HandleCallOperation, (operationLeft, operationRight) => {
    const mapOperation = new Map();
    const mapOperationRight = operationRight.getMapOperation()

    for (const [publicKey, status] of operationLeft.getMapOperation().entries()) {
      const statusRight = mapOperationRight.get(publicKey);
      mapOperation.set(publicKey, {
        prev: status.prev,
        next: statusRight.next
      });
    }

    return new HandleCallOperation(mapOperation);
  })
  .withSquashFunction(DropCallOperation, HandleCallOperation, doSquash((operationLeft, operationRight) => {
    if (operationLeft.isInvert()) {
      const handled = operationLeft.getHandled();

      for (const [publicKey, status] of operationRight.getMapOperation().entries()) {
        if (status.next === null) {
          handled.delete(publicKey);
        } else {
          handled.set(publicKey, status.next);
        }
      }

      return new DropCallOperation(
        operationLeft.getPubKey(),
        operationLeft.getPeerId(),
        operationLeft.getTimestamp(),
        handled,
        operationLeft.getDropTime(),
        true
      );
    }

    return null;
  }))
  .withSquashFunction(HandleCallOperation, DropCallOperation, doSquash((operationLeft, operationRight) => {
    if (!operationRight.isInvert()) {
      const handled = operationRight.getHandled();

      for (const [publicKey, status] of operationLeft.getMapOperation().entries()) {
        if (status.next === null) {
          handled.set(publicKey, status.prev);
        } else {
          handled.delete(publicKey);
        }
      }

      return DropCallOperation.create(
        operationRight.getPubKey(),
        operationRight.getPeerId(),
        operationRight.getTimestamp(),
        handled,
        operationRight.getDropTime()
      );
    }

    return null;
  }))
  .build();

function doSquash(squashFn) {
  return (operationLeft, operationRight) => {
    if (operationLeft.isEmpty()) {
      return operationRight;
    }

    if (operationRight.isEmpty()) {
      return operationLeft;
    }

    return squashFn(operationLeft, operationRight);
  }
}

export default chatRoomOTSystem;

File: global-apps/global-chat/src/main/java/io/global/chat/chatroom/ChatRoomOTSystem.java
package io.global.chat.chatroom;

import io.datakernel.ot.OTSystem;
import io.datakernel.ot.OTSystemImpl;
import io.datakernel.ot.OTSystemImpl.SquashFunction;
import io.datakernel.ot.OTSystemImpl.TransformFunction;
import io.datakernel.ot.TransformResult;
import io.datakernel.ot.exceptions.OTTransformException;
import io.global.chat.chatroom.operation.*;
import io.global.common.PubKey;
import io.global.ot.map.MapOperation;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.function.Function;

import static io.datakernel.common.collection.CollectionUtils.first;
import static io.global.chat.Utils.HANDLE_CALL_SUBSYSTEM;
import static io.global.chat.chatroom.operation.DropCallOperation.dropCall;
import static java.util.Arrays.asList;
import static java.util.Collections.singletonList;
import static java.util.stream.Collectors.toList;

public final class ChatRoomOTSystem {
	private ChatRoomOTSystem() {
		throw new AssertionError();
	}

	public static OTSystem<ChatRoomOperation> createOTSystem() {

		return OTSystemImpl.<ChatRoomOperation>create()
				.withEmptyPredicate(MessageOperation.class, MessageOperation::isEmpty)
				.withEmptyPredicate(CallOperation.class, CallOperation::isEmpty)
				.withEmptyPredicate(DropCallOperation.class, DropCallOperation::isEmpty)
				.withEmptyPredicate(HandleCallOperation.class, HandleCallOperation::isEmpty)

				.withInvertFunction(MessageOperation.class, op -> singletonList(op.invert()))
				.withInvertFunction(CallOperation.class, op -> singletonList(op.invert()))
				.withInvertFunction(DropCallOperation.class, op -> singletonList(op.invert()))
				.withInvertFunction(HandleCallOperation.class, op -> singletonList(op.invert()))

				.withTransformFunction(MessageOperation.class, MessageOperation.class, swap())
				.withTransformFunction(MessageOperation.class, CallOperation.class, swap())
				.withTransformFunction(MessageOperation.class, DropCallOperation.class, swap())
				.withTransformFunction(MessageOperation.class, HandleCallOperation.class, swap())

				.withTransformFunction(CallOperation.class, CallOperation.class, (left, right) -> {
					if (!Objects.equals(left.getPrev(), right.getPrev())) {
						throw new OTTransformException("Previous values should be equal");
					}

					CallInfo leftNextValue = left.getNext();
					CallInfo rightNextValue = right.getNext();

					if (Objects.equals(leftNextValue, rightNextValue)) {
						return TransformResult.empty();
					}

					if (leftNextValue == null) {
						return TransformResult.left(asList(left.invert(), right));
					}
					if (rightNextValue == null) {
						return TransformResult.right(asList(right.invert(), left));
					}
					if (leftNextValue.getCallStarted() >= rightNextValue.getCallStarted()) {
						return TransformResult.right(asList(right.invert(), left));
					} else {
						return TransformResult.left(asList(left.invert(), right));
					}
				})
				.withTransformFunction(CallOperation.class, DropCallOperation.class, (left, right) -> {
					if (left.getNext() != null) {
						return TransformResult.right(asList(right.invert(), left));
					}
					if (!right.isInvert()) {
						return TransformResult.left(asList(left.invert(), right));
					}
					throw new OTTransformException();
				})
				.withTransformFunction(CallOperation.class, HandleCallOperation.class, (left, right) -> {
					if (left.getNext() == null) {
						return TransformResult.right(asList(right.invert(), left));
					}
					return TransformResult.of(right, left);
				})

				.withTransformFunction(DropCallOperation.class, DropCallOperation.class, (left, right) -> {
					if (left.equals(right)) {
						return TransformResult.empty();
					}
					if (!left.isInvert() && !right.isInvert()) {
						if (left.getDropTimestamp() <= right.getDropTimestamp()) {
							return TransformResult.right(asList(right.invert(), left));
						} else {
							return TransformResult.left(asList(left.invert(), right));
						}
					}
					throw new OTTransformException();
				})
				.withTransformFunction(DropCallOperation.class, HandleCallOperation.class, (left, right) -> {
					if (!left.isInvert()) {
						CallInfo callInfo = left.getCallInfo();
						Map<PubKey, Boolean> previouslyHandled = new HashMap<>(left.getPreviouslyHandled());

						right.getMapOperation().getOperations()
								.forEach((user, setValue) -> {
									Boolean next = setValue.getNext();
									if (next == null) {
										previouslyHandled.remove(user);
									} else {
										previouslyHandled.put(user, next);
									}
								});
						long dropTimestamp = left.getDropTimestamp();
						return TransformResult.right(dropCall(callInfo, previouslyHandled, dropTimestamp));
					}
					throw new OTTransformException();
				})

				.withTransformFunction(HandleCallOperation.class, HandleCallOperation.class, (left, right) -> {
					TransformResult<MapOperation<PubKey, Boolean>> transform = HANDLE_CALL_SUBSYSTEM
							.transform(left.getMapOperation(), right.getMapOperation());
					return TransformResult.of(
							collect(transform.left, HandleCallOperation::new),
							collect(transform.right, HandleCallOperation::new)
					);
				})

				.withSquashFunction(MessageOperation.class, MessageOperation.class, doSquash((first, second) -> {
					if (first.isInversionFor(second)) {
						return MessageOperation.EMPTY;
					}
					return null;
				}))
				.withSquashFunction(CallOperation.class, CallOperation.class, doSquash((first, second) -> {
					if (first.getPrev() != null && first.getNext() != null && second.getPrev() != null && second.getNext() != null) {
						return new CallOperation(first.getPrev(), second.getNext());
					}
					return null;
				}))
				.withSquashFunction(DropCallOperation.class, DropCallOperation.class, doSquash((first, second) -> {
					if (first.isInversionFor(second)) {
						return DropCallOperation.EMPTY;
					}
					return null;
				}))
				.withSquashFunction(HandleCallOperation.class, HandleCallOperation.class, (first, second) -> {
					List<MapOperation<PubKey, Boolean>> toBeSquashed = asList(first.getMapOperation(), second.getMapOperation());
					List<MapOperation<PubKey, Boolean>> squashed = HANDLE_CALL_SUBSYSTEM.squash(toBeSquashed);
					return new HandleCallOperation(first(squashed));
				})
				.withSquashFunction(DropCallOperation.class, HandleCallOperation.class, doSquash((first, second) -> {
					if (first.isInvert()) {
						Map<PubKey, Boolean> previouslyHandled = new HashMap<>(first.getPreviouslyHandled());

						second.getMapOperation().getOperations()
								.forEach((user, setValue) -> {
									Boolean next = setValue.getNext();
									if (next == null) {
										previouslyHandled.remove(user);
									} else {
										previouslyHandled.put(user, next);
									}
								});
						return new DropCallOperation(first.getCallInfo(), previouslyHandled, first.getDropTimestamp(), true);
					}
					return null;
				}))
				.withSquashFunction(HandleCallOperation.class, DropCallOperation.class, doSquash((first, second) -> {
					if (!second.isInvert()) {
						Map<PubKey, Boolean> previouslyHandled = new HashMap<>(second.getPreviouslyHandled());

						first.getMapOperation().getOperations()
								.forEach((user, setValue) -> {
									if (setValue.getNext() == null) {
										previouslyHandled.put(user, setValue.getPrev());
									} else {
										previouslyHandled.remove(user);
									}
								});
						return new DropCallOperation(second.getCallInfo(), previouslyHandled, second.getDropTimestamp(), false);
					}
					return null;
				}));
	}

	private static <L extends ChatRoomOperation, R extends ChatRoomOperation> TransformFunction<ChatRoomOperation, L, R> swap() {
		return (left, right) -> {
			if (left.equals(right)) {
				return TransformResult.empty();
			}
			return TransformResult.of(right, left);
		};
	}

	private static <O1 extends ChatRoomOperation, O2 extends ChatRoomOperation> SquashFunction<ChatRoomOperation, O1, O2> doSquash(SquashFunction<ChatRoomOperation, O1, O2> squashFn) {
		return (first, second) -> {
			if (first.isEmpty()) return second;
			if (second.isEmpty()) return first;
			return squashFn.trySquash(first, second);
		};
	}

	private static <O, S> List<O> collect(List<S> ops, Function<S, O> constructor) {
		return ops.stream().map(constructor).collect(toList());
	}
}
